engine/src/main/scala/cromwell/engine/workflow/JobInputEvaluator.scala

```diff
@@ -0,0 +1,176 @@
+package cromwell.engine.workflow
+
+import cromwell.backend.BackendJobDescriptorKey
+import cromwell.engine.EngineWorkflowDescriptor
+import cromwell.engine.ExecutionIndex._
+import cromwell.engine.workflow.lifecycle.WorkflowExecutionActor.{OutputCallKey, OutputEntry, OutputStore}
+import cromwell.util.TryUtil
+import wdl4s._
+import wdl4s.expression.WdlStandardLibraryFunctions
+import wdl4s.values.{WdlCallOutputsObject, WdlValue}
+
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+/**
+   * Resolve and evaluate call input declarations
+   *
+   * As a result, the following wdl
+   *
+   * task t {
+   *   File q
+   *   String r = read_string(q) #(1)
+   *   String s
+   * }
+   *
+   * workflow w {
+   *   File f
+   *   String a = read_string(f) #(2)
+   *   call t { input:
+   *              s = a + read_string(f), #(3)
+   *              q = f
+   *          }
+   * }
+   *
+   * will be evaluated such that
+   *
+   * read_string (1) will be evaluated using BFN
+   * read_string (2) will be evaluated using EFN
+   * read_string (3) will be evaluated using BFN
+   *
+   * Where
+   *   BFN is the WDL function implementation from the backend t will run on
+   *   EFN is the WDL function implementation from the engine
+   *
+   * q, r and s will be evaluated sequentially, such that when evaluating r,
+   * the value of q has already been resolved in the appropriate scope and evaluated by the appropriate wdl functions.
+   */
+class JobInputEvaluator(workflowDescriptor: EngineWorkflowDescriptor) {
+
+  private val calls = workflowDescriptor.backendDescriptor.workflowNamespace.workflow.calls
+
+  /*
+   *  /|\ this could be very time consuming, as it potentially involves IO when using engine functions
+   * /_._\
+   */
```

Comment:
I think any sentence to qualify this with the word "art" in it is most likely far from reality